modules/cudafilters/test/test_filters.cpp

```diff
@@ -647,4 +648,57 @@ INSTANTIATE_TEST_CASE_P(CUDA_Filters, MorphEx, testing::Combine(
     testing::Values(Iterations(1), Iterations(2), Iterations(3)),
     WHOLE_SUBMAT));
 
+/////////////////////////////////////////////////////////////////////////////////////////////////
+// Median
+
+
+PARAM_TEST_CASE(Median, cv::cuda::DeviceInfo, cv::Size, MatDepth,  KernelSize, UseRoi)
+{
+    cv::cuda::DeviceInfo devInfo;
+    cv::Size size;
+    int type;
+    int kernel;
+    bool useRoi;
+
+    virtual void SetUp()
+    {
+        devInfo = GET_PARAM(0);
+        size = GET_PARAM(1);
+        type = GET_PARAM(2);
+        kernel = GET_PARAM(3);
+        useRoi = GET_PARAM(4);
+
+        cv::cuda::setDevice(devInfo.deviceID());
+    }
+};
+
+CUDA_TEST_P(Median, Accuracy)
+{
+    cv::Mat src = randomMat(size, type);
+
+    cv::Ptr<cv::cuda::Filter> median = cv::cuda::createMedianFilter(src.type(), kernel);
+
+    cv::cuda::GpuMat dst = createMat(size, type, useRoi);
+    median->apply(loadMat(src, useRoi), dst);
+
+    cv::Mat dst_gold;
+    cv::medianBlur(src,dst_gold,kernel);
+
+    EXPECT_MAT_SIMILAR(dst_gold, dst, 1.0);
```

Comment:
I have made all the requested changes.

Let me explain to you why there is some difference in the outputs of the algorithms.

This new CUDA based algorithm does one key thing different in comparison with the its CPU counterpart. The CUDA version does not create "a border" around the image for the kernel movement purposes. This results in different values for pixels on the borders of the image.

Below is a comparison of the CPU implementation with the CUDA implementation for the "Barbara" image - this is not a random matrix. I am using a square BW 512x512 version of Barbara. Grid-size is a "performance" tuning parameter used to state how many partitions the image should be divided into. As you can see the PSNR value comparing these implementations is more than decent.


system,image-size,kernel-size,time,grid-size
GPU,512,3,0.010942,128
PSNR :40.0033
GPU,512,5,0.010370,128
PSNR :38.6299
GPU,512,7,0.010846,128
PSNR :43.2202
GPU,512,9,0.010812,128
PSNR :42.2529
GPU,512,11,0.011241,128
PSNR :44.4187
GPU,512,13,0.011194,128
PSNR :43.389
GPU,512,15,0.011630,128
PSNR :46.3429





________________________________
From: Vladislav Vinogradov <notifications@github.com>
Sent: Monday, January 11, 2016 2:26 AM
To: Itseez/opencv
Cc: Green, Oded
Subject: Re: [opencv] Implemented median filter in CUDA (#5888)


In modules/cudafilters/test/test_filters.cpp<https://github.com/Itseez/opencv/pull/5888#discussion_r49294498>:

[https://avatars1.githubusercontent.com/u/1091858?v=3&s=400]<https://github.com/Itseez/opencv/pull/5888#discussion_r49294498>

Implemented median filter in CUDA by ogreen · Pull Request ...<https://github.com/Itseez/opencv/pull/5888#discussion_r49294498>
github.com
Implemented median filter in CUDA using the histogram-based operation formulation from the paper " Median Filtering in Constant Time". The CPU implementation in ...



> +    }
> +};
> +
> +CUDA_TEST_P(Median, Accuracy)
> +{
> +    cv::Mat src = randomMat(size, type);
> +
> +    cv::Ptr<cv::cuda::Filter> median = cv::cuda::createMedianFilter(src.type(), kernel);
> +
> +    cv::cuda::GpuMat dst = createMat(size, type, useRoi);
> +    median->apply(loadMat(src, useRoi), dst);
> +
> +    cv::Mat dst_gold;
> +    cv::medianBlur(src,dst_gold,kernel);
> +
> +    EXPECT_MAT_SIMILAR(dst_gold, dst, 1.0);


1.0 is a max threshold for EXPECT_MAT_SIMILAR, which means that results doesn't match. Could you check why the CUDA results doesn't match with CPU.

-
Reply to this email directly or view it on GitHub<https://github.com/Itseez/opencv/pull/5888/files#r49294498>.